### ISSUE TITLE
Async functions

### ISSUE THREAD

**Issue by callmeahab** [georust/gdal#104]
Hi,

Is there a plan to move functions to support async execution, or at least implement thread safe data types, while this crate is good, its really hard to implement functions that require async operations in conjunction with gdal functions.

Thanks

**Comment by jdroenner** [comment]
Since GDAL does not really support async operations i guess this is out of scope for this crate. The only async feature in GDAL i know of is the `GDALAsyncReader `. It would be great to support this, however as far as i understand the async read calls a blocking read in most drivers.

For thread safe types we can only support what GDAL provides. We added `Send` for datasets since GDAL states they might be used from different threads. However, i'm not sure what else is possible. If you have suggestions don't hesitate to share them.

**Comment by callmeahab** [comment]
I posted this issue because I was having trouble running a background worker using an actix actor, compiler was complaining about all the types that don't implement Send trait once I tried using Arbiter to create background thread. However it worked when using actix_rt::spawn with an async function. In general for my use performance is secondary, I just wanted to create a background thread so user doesn't have to wait around.

As far as I understand this is mostly a wrapper for gdal's c api, but are there plans to write these functions natively in rust? In that case, if time permits I might be willing to help, but would require guidance.

**Comment by lnicola** [comment]
GDAL is a old, but still actively-maintained project. It makes sense to have a nice set of bindings to it, but probably less so to RIIR. If would be impossible to keep up with it, not to mention that the IO drivers themselves (e.g. `libtiff`) are already using synchronous file operations. For Linux, there wasn't even a good async IO file solution until very recently, and GDAL supports dozens of formats on multiple platforms.

So using a thread pool is probably the best solution. Making `Dataset` `Send` probably helped, but it's not released to crates.io yet, so you should try using a `git` dependency.

**Comment by callmeahab** [comment]
Ok, makes sense. Thanks for your assistance.

**Comment by lnicola** [comment]
If you're doing `actix-web` and still have trouble, I have a small app that serves tiles from a raster, ping me and I can share it.

**Comment by callmeahab** [comment]
I resolved the original issue, which was how to run background task with an async function which uses gdal. When using actors I got errors that gdal functions don't implement `Send`, however `actix_rt::spawn` works fine.
Do you know if its possible to use rayon with gdal functions?

**Comment by lnicola** [comment]
I don't think so (`Dataset` is `!Sync`), however the GDAL docs say:

> The returned dataset should only be accessed by one thread at a time. If you want to use it from different threads, you must add all necessary code (mutexes, etc.) to avoid concurrent use of the object. (Some drivers, such as GeoTIFF, maintain internal state variables that are updated each time a new block is read, thus preventing concurrent use.)

So you would probably need to open the dataset multiple times in order to read in parallel from it.

**Comment by callmeahab** [comment]
Ok, thanks
